File: Easy_Leetcode_Problems/Easy_Leetcode_Problems/Contains_duplicate_217.java
// 💡 Logic
//
// We will use a HashSet to find duplicate elements.
//
// A HashSet does not store duplicate values.
//
// For example:
//
// nums = [1, 2, 3, 1]
//
// HashSet = {1, 2, 3}
//
// Original array length = 4
// HashSet length = 3
//
// Since both lengths are different, a duplicate exists.
//
// Therefore:
//
// nums.length != set.size()
//
// If this condition is true, the array contains a duplicate.
//
//
// 🔄 Algorithm
//
// 1. Create a HashSet.
//
// 2. Add every element of nums into the HashSet.
//
// 3. If an element is already present in the HashSet,
//    return true because a duplicate is found.
//
// 4. If all elements are added successfully,
//    return false because all elements are distinct.
//
//
// 🧑‍💻 Java Code

import java.util.*;

class Solution {
    public boolean containsDuplicate(int[] nums) {

        // Create a HashSet to store unique elements
        HashSet<Integer> set = new HashSet<>();

        // Traverse through the array
        for (int num : nums) {

            // If the element already exists in the set,
            // a duplicate is found
            if (set.contains(num)) {
                return true;
            }

            // Add the element to the set
            set.add(num);
        }

        // No duplicate was found
        return false;
    }
}


// ⭐ Shortest Code

class Solution {
    public boolean containsDuplicate(int[] nums) {

        // Create a HashSet containing all elements
        Set<Integer> set = new HashSet<>();

        // Add all elements to the set
        for (int num : nums) {
            set.add(num);
        }

        // Different sizes mean duplicates exist
        return nums.length != set.size();
    }
}


// 🧪 Example
//
// nums = [1, 2, 3, 1]
//
// Initially:
//
// set = {}
//
// Add 1:
// set = {1}
//
// Add 2:
// set = {1, 2}
//
// Add 3:
// set = {1, 2, 3}
//
// Next element is 1.
//
// 1 is already present in the set.
//
// Therefore:
//
// return true
//
//
// ⏱️ Complexity
//
// Time Complexity: O(n)
//
// Space Complexity: O(n)
//
// n = number of elements in the array.
//
//
// 🎯 Interview Explanation
//
// "I use a HashSet to detect duplicate elements.
// HashSet stores only unique values.
// While traversing the array, I check whether the current
// element is already present in the HashSet.
// If it is present, I immediately return true.
// If I finish traversing the array without finding a duplicate,
// I return false."

File: Easy_Leetcode_Problems/Easy_Leetcode_Problems/Contains_duplicate_217.py
# 💡 Logic
#
# We will use a Set to find duplicate elements.
#
# A Set does not store duplicate values.
#
# For example:
#
# nums = [1, 2, 3, 1]
#
# set(nums) = {1, 2, 3}
#
# len(nums) = 4
# len(set(nums)) = 3
#
# Since both lengths are different, a duplicate exists.
#
# Therefore:
#
# len(nums) != len(set(nums))
#
# If this condition is True, the array contains a duplicate.
#
#
# 🔄 Algorithm
#
# 1. Convert nums into a Set.
#
# 2. Compare the length of nums with the length of the Set.
#
# 3. If both lengths are different:
#    → Return True because a duplicate exists.
#
# 4. If both lengths are the same:
#    → Return False because all elements are distinct.
#
#
# 🧑‍💻 Python Code
#
# class Solution:
#     def containsDuplicate(self, nums: list[int]) -> bool:
#
#         # Convert the array into a set
#         unique = set(nums)
#
#         # If lengths are different, duplicates exist
#         if len(nums) != len(unique):
#             return True
#
#         # All elements are distinct
#         return False
#
#
# ⭐ Shortest Code
#
# class Solution:
#     def containsDuplicate(self, nums: list[int]) -> bool:
#         return len(nums) != len(set(nums))
#
#
# 🧪 Example
#
# nums = [1, 2, 3, 1]
#
# Original array:
# [1, 2, 3, 1]
#
# Set:
# {1, 2, 3}
#
# Length of nums = 4
# Length of set = 3
#
# 4 != 3 → True
#
# Therefore, a duplicate exists.
#
#
# ⏱️ Complexity
#
# Time Complexity: O(n)
#
# Space Complexity: O(n)
#
# n = number of elements in the array.
#
#
# 🎯 Interview Explanation
#
# "I use a Set to detect duplicate elements.
# A Set stores only unique values, so if the length of the
# original array is different from the length of the Set,
# it means that at least one element appeared more than once.
# Therefore, I return True when the lengths are different;
# otherwise, I return False."